xds/internal/test/e2e/e2e_test.go

```diff
@@ -0,0 +1,257 @@
+		DialTarget: testName,
+		NodeID:     cp.nodeID,
+		Host:       "localhost",
+		Port:       serverPort,
+		SecLevel:   e2e.SecurityLevelNone,
+	})
+
+	// Update EDS to multiple backends.
+	var ports []uint32
+	for _, s := range servers {
+		ports = append(ports, uint32(s.port))
+	}
+	edsMsg := resources.Endpoints[0]
+	resources.Endpoints[0] = e2e.DefaultEndpoint(
+		edsMsg.ClusterName,
+		"localhost",
+		ports,
+	)
+
+	// Update CDS lbpolicy to ringhash.
+	cdsMsg := resources.Clusters[0]
+	cdsMsg.LbPolicy = v3clusterpb.Cluster_RING_HASH
+
+	// Update RDS to hash the header.
+	rdsMsg := resources.Routes[0]
+	rdsMsg.VirtualHosts[0].Routes[0].Action = &v3routepb.Route_Route{Route: &v3routepb.RouteAction{
+		ClusterSpecifier: &v3routepb.RouteAction_Cluster{Cluster: cdsMsg.Name},
+		HashPolicy: []*v3routepb.RouteAction_HashPolicy{{
+			PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+				Header: &v3routepb.RouteAction_HashPolicy_Header{
+					HeaderName: testMDKey,
+				},
+			},
+		}},
+	}}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	if err := cp.server.Update(ctx, resources); err != nil {
+		t.Fatalf("failed to update control plane resources: %v", err)
+	}
+
+	// We can skip CSDS check because there's no long delay as in TD.
+
+	st, err := c.clientStats(ctx)
+	if err != nil {
+		t.Fatalf("failed to get client stats: %v", err)
+	}
+	if st.NumFailures != 0 {
+		t.Fatalf("Got %v failures: %+v", st.NumFailures, st)
+	}
+	if len(st.RpcsByPeer) != 1 {
+		t.Fatalf("more than 1 backends got traffic: %v, want 1", st.RpcsByPeer)
+	}
+
+	// Call channelz to verify that only one subchannel is in state Ready.
+	scs, err := c.channelzSubChannels(ctx)
+	if err != nil {
+		t.Fatalf("failed to fetch channelz: %v", err)
+	}
+	verifySubConnStates(t, scs, map[channelzpb.ChannelConnectivityState_State]int{
+		channelzpb.ChannelConnectivityState_READY: 1,
+		channelzpb.ChannelConnectivityState_IDLE:  2,
+	})
+
+	// Send Unary call with different metadata value with integers starting from
+	// 0. Stop when a second peer is picked.
+	var (
+		diffPeerPicked bool
+		mdValue        int
+	)
+	for !diffPeerPicked {
+		if err := c.configRPCs(ctx, &testpb.ClientConfigureRequest{
+			Types: []testpb.ClientConfigureRequest_RpcType{
+				testpb.ClientConfigureRequest_EMPTY_CALL,
+				testpb.ClientConfigureRequest_UNARY_CALL,
+			},
+			Metadata: []*testpb.ClientConfigureRequest_Metadata{
+				{Type: testpb.ClientConfigureRequest_EMPTY_CALL, Key: testMDKey, Value: testMDValue},
+				{Type: testpb.ClientConfigureRequest_UNARY_CALL, Key: testMDKey, Value: strconv.Itoa(mdValue)},
+			},
+		}); err != nil {
+			t.Fatalf("failed to configure RPC: %v", err)
+		}
+
+		st, err := c.clientStats(ctx)
+		if err != nil {
+			t.Fatalf("failed to get client stats: %v", err)
+		}
+		if st.NumFailures != 0 {
+			t.Fatalf("Got %v failures: %+v", st.NumFailures, st)
+		}
+		if len(st.RpcsByPeer) == 2 {
+			break
+		}
```

Comment:
this block of code seems like would be repeated in every test. Does it make sense to make a helper?

Reply:
This check won't repeat a lot. Different tests check different things.
This affinity test check number of peers. Routing tests (not available yet) will check RPC distribution.